File: frontend/src/types/quote.ts
export type QuoteStatus = 'BORRADOR' | 'ENVIADA' | 'APROBADA' | 'RECHAZADA';

export const QUOTE_STATUS_LABELS: Record<QuoteStatus, string> = {
  BORRADOR: 'Borrador',
  ENVIADA: 'Enviada',
  APROBADA: 'Aprobada',
  RECHAZADA: 'Rechazada',
};

export const QUOTE_STATUS_COLORS: Record<QuoteStatus, string> = {
  BORRADOR: 'bg-gray-100 text-gray-700',
  ENVIADA: 'bg-blue-100 text-blue-700',
  APROBADA: 'bg-green-100 text-green-700',
  RECHAZADA: 'bg-red-100 text-red-700',
};

export interface QuoteItem {
  id: number | null;
  descripcion: string;
  cantidad: number;
  precioUnitario: number;
  subtotal: number;
  orden: number;
}

export interface QuoteItemRequest {
  descripcion: string;
  cantidad: number;
  precioUnitario: number;
}

export interface Quote {
  id: number;
  numero: string;
  customerId: number;
  customerNombre: string;
  assetId: number | null;
  assetDescripcion: string | null;
  estado: QuoteStatus;
  notas: string | null;
  total: number;
  items: QuoteItem[];
  createdAt: string;
  updatedAt: string;
}

export interface QuoteRequest {
  customerId: number;
  assetId?: number | null;
  notas?: string | null;
  items: QuoteItemRequest[];
}

File: src/main/java/cotizaciones/app/Model/Quote.java
package cotizaciones.app.Model;



import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "quotes")
@Getter
@Setter
public class Quote {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true, length = 30)
    private String numero;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "customer_id", nullable = false)
    private Customer customer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "asset_id")
    private Asset asset;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private QuoteStatus estado = QuoteStatus.BORRADOR;

    @Column(columnDefinition = "text")
    private String notas;

    @Column(nullable = false, precision = 14, scale = 2)
    private BigDecimal total = BigDecimal.ZERO;

    @OneToMany(
            mappedBy = "quote",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    @OrderBy("orden ASC")
    private List<QuoteItem> items = new ArrayList<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @PrePersist
    void onCreate() {
        var now = OffsetDateTime.now();
        createdAt = now;
        updatedAt = now;
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }

    /**
     * Agrega un ítem y establece la relación bidireccional.
     */
    public void addItem(QuoteItem item) {
        items.add(item);
        item.setQuote(this);
    }

    /**
     * Reemplaza todos los ítems. Con orphanRemoval, los viejos se borran solos.
     */
    public void replaceItems(List<QuoteItem> nuevos) {
        items.clear();
        for (QuoteItem item : nuevos) {
            addItem(item);
        }
    }

    /**
     * Recalcula el total sumando los subtotales de todos los ítems.
     */
    public void recalcularTotal() {
        this.total = items.stream()
                .map(QuoteItem::getSubtotal)
                .filter(java.util.Objects::nonNull)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
}